### ISSUE TITLE
Cannot infer type of parameterized function declared as class property

### ISSUE THREAD

**Issue by zgotsch** [facebook/flow#5193]
In the docs (https://flow.org/en/docs/types/classes/#toc-class-fields-properties), it's indicated that the type of class properties can be inferred. However, when using a function with a class type parameter in the arguments, the type must be explicitly indicated to avoid an error:

```
This type is incompatible with the expected param type of some incompatible instantiation of...
```

https://flow.org/try/#0MYGwhgzhAECyCeBhcUA8AVAfNA3gKGmgHojp0ALASxgHcB7AJwGsIAuA40ygO24FMGraAAoAHkPQBKaAF5sAIzp0QfMN1kjxZSUMXLV6ubg4kyVGGBAQ60eszYmuvAWIk7oelWuOEe-Bhqu2rpKXobY+ISEDHwALgCuDOqxDPF8ANwcAL54OXjAdNwQsdA8xRr8NHBIKBDCkpllsQB0fi4ALABMDUA

The property typechecks if the type is manually annotated. Additionally, the error only occurs if there is an invocation of the property function, which seems wrong.

**Comment by bradennapier** [comment]
Yeah --- I can not make sense of this... been at this for hours now... I spent the last 5 days dealing with this type of stuff and somehow gt a lot of it working (or so I think I do)... but for the life of me I can't see how this could possibly not be working:

```
class MyClass<T> {
  inner = (x: T): T => {
    return x;
  }
}

const inst: MyClass<number> = new MyClass();
inst.inner(42);
```

```
8: inst.inner(42);
              ^ number. This type is incompatible with the expected param type of
2:   inner = (x: T): T => {
                 ^ some incompatible instantiation of `T`
```

Thought I had this all figured out and even started to write a tutorial on these things thinking I understand it all and sure enough cant even make the simplest examples work still after 12 weeks of working with Flow....

Simple stuff is obviously simple to do - but the second you work with interfaces or generics things definitely do not work right.

**Comment by ddgromit** [comment]
Looks a dupe of #4743

**Comment by rpominov** [comment]
I have a similar issue. Seems like there is bug related to `foo = (): T => { ... }` syntax.

```js
// @flow

class Foo<T> {
  
  // doesn't work
  foo = (): T => {
    return (null: any)
  }

  // works
  bar: (() => T) = () => {
    return (null: any)
  }

  baz(): void {
    // error
    ;(this.foo(): T)
    
    // ok
    ;(this.bar(): T)
  }
}

```

https://flow.org/try/#0PTAEAEDMBsHsHcBQiDG0CGBnToBitYAeAFQD5QBvRUUa0EUAE1gFNMA7AcgBdR5YATgGs6kAqAC8oABQBKAFyhik8lRo0BLbgFcB7Ge23Roi9OwCesugF9kNBv2GY6AI3QDF0uSqWzJMvwlVOg0tXX1pQ2NTCysaW1d0AC85RQA3WABLRkoQ+jAWAQFBPIBuaW4AC0zMADoxWFTfPLyGWBF1UHKqmtq3ASbiONBbWyA

**Comment by deecewan** [comment]
Seems related to https://github.com/facebook/flow/issues/6401 - there are problems with generics inside instance properties.

**Comment by SamChou19815** [comment]
Now explicit type annotation on new call is required, and it works now

**Comment by bradennapier** [comment]
Yay!  Good work!  Great to see the effort to maintain old issues like this with updates on status -- evne if it took 7 years ;-)
